kimchi/src/folding/error.rs

```diff
@@ -0,0 +1,331 @@
+use crate::folding::{
+    expressions::{Degree, ExtendedFoldingColumn, FoldingExp, IntegratedFoldingExpr},
+    quadricization::ExtendedWitnessGenerator,
+    EvalLeaf, FoldingConfig, FoldingEnv, RelaxedInstance, RelaxedWitness,
+};
+use ark_ec::AffineCurve;
+use ark_ff::Field;
+use ark_poly::{Evaluations, Radix2EvaluationDomain};
+use poly_commitment::SRS;
+
+#[derive(Clone, Copy)]
+pub enum Side {
+    Left,
+    Right,
+}
+impl std::ops::Not for Side {
+    type Output = Self;
+
+    fn not(self) -> Self::Output {
+        match self {
+            Side::Left => Side::Right,
+            Side::Right => Side::Left,
+        }
+    }
+}
+type Fi<C> = <<C as FoldingConfig>::Curve as AffineCurve>::ScalarField;
+
+///evaluates the expression in the provided side
+pub(crate) fn eval_sided<'a, C: FoldingConfig>(
+    exp: &FoldingExp<C>,
+    env: &'a ExtendedEnv<C>,
+    side: Side,
+) -> EvalLeaf<'a, Fi<C>> {
+    use FoldingExp::*;
+
+    let eval_bin = |e1, e2| (eval_exp_error(e1, env, side), eval_exp_error(e2, env, side));
+    match exp {
+        Cell(col) => env.col(col, side),
+        Double(e) => {
+            let col = eval_exp_error(e, env, side);
+            col.map(Field::double, |f| {
+                Field::double_in_place(f);
+            })
+        }
+        Square(e) => {
+            let col = eval_exp_error(e, env, side);
+            col.map(Field::square, |f| {
+                Field::square_in_place(f);
+            })
+        }
+        Add(e1, e2) => {
+            let (a, b) = eval_bin(e1, e2);
+            EvalLeaf::bin_op(a, b, |a, b| *a + b, |a, b| *a += b)
+        }
+        Sub(e1, e2) => {
+            let (a, b) = eval_bin(e1, e2);
+            EvalLeaf::bin_op(a, b, |a, b| *a - b, |a, b| *a -= b)
+        }
+        Mul(e1, e2) => {
+            let (a, b) = eval_bin(e1, e2);
+            EvalLeaf::bin_op(a, b, |a, b| *a * b, |a, b| *a *= b)
+        }
+    }
+}
+pub(crate) fn eval_exp_error<'a, C: FoldingConfig>(
+    exp: &FoldingExp<C>,
+    env: &'a ExtendedEnv<C>,
+    side: Side,
+) -> EvalLeaf<'a, Fi<C>> {
+    let degree = exp.folding_degree();
```

Comment:
This isn't used for most of the operations, it would be more efficient to compute this where it is used.